gson/src/org/immutables/gson/stream/JsonGeneratorWriter.java

```diff
@@ -145,7 +150,25 @@ public JsonWriter value(Number value) throws IOException {
         throw new IllegalArgumentException("JSON forbids NaN and infinities: " + value);
       }
     }
-    generator.writeNumber(d);
+    if (value instanceof Integer) {
+      generator.writeNumber((Integer) value);
+    } else if (value instanceof Short) {
+      generator.writeNumber((Short) value);
+    } else if (value instanceof Long) {
+      generator.writeNumber((Long) value);
+    } else if (value instanceof Float) {
+      generator.writeNumber((Float) value);
+    } else if (value instanceof Double) {
+      generator.writeNumber((Double) value);
+    } else if (value instanceof BigInteger) {
+      generator.writeNumber((BigInteger) value);
+    } else if (value instanceof BigDecimal) {
+      generator.writeNumber((BigDecimal) value);
+    } else if (floor(d) == d) {
```

Comment:
Please move the [statement](https://github.com/immutables/immutables/blob/9d6392258ce5b47b09d7d6b35a5f5b168de3471b/gson/src/org/immutables/gson/stream/JsonGeneratorWriter.java#L142) `double d = value.doubleValue()` (not visible in this review) to where conversion to double is necessary. For instance you don't need conversion to double when you know that current type is `Integer`.

Reply:
What about the strict validation you were performing?

Reply:
Left the `double d = value.doubleValue()` in there to enable isNan strict checks